### ISSUE TITLE
prost automatic naming convention causes build failure on valid protobuf

### ISSUE THREAD

**Issue by gen-xu** [tokio-rs/prost#665]
Minimum re-produceable example of protobuf
```protobuf
syntax = "proto3";
package foobar;
enum Fb { 
    Fb1 = 0;
}
enum FB {
    FB2 = 0;
}

```
Since prost rename `FB` into `Fb` and it generates conflicts definitions like following and causes compiling failure in some cases

Is there any way to enforce naming convention in prost? I think it relates with this issue https://github.com/tokio-rs/prost/issues/593

But since this causes compiling failure only in prost but works with other languages with protoc I would deem this as a bug.

```rust
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, ::prost::Enumeration)]
#[repr(i32)]
pub enum Fb {
    Fb1 = 0,
}
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, ::prost::Enumeration)]
#[repr(i32)]
pub enum Fb {
    Fb2 = 0,
}

```

**Comment by gen-xu** [comment]
@LucioFranco yes I agree it goes against naming convention. But sometime I think there are needs for that. For example XML, GPIO are very commonly used abbreviations together as all capitals cases. Also I am not sure if this is a bug of heck that does the naming conversion in Prost. As to me I think the upper camel cases applied on FB should still lead into FB.

**Comment by BatmanAoD** [comment]
Would it be reasonable to support suppressing renaming for `message`s and `enum`s entirely?

That seems similar to what's requested [here](https://github.com/tokio-rs/prost/issues/207), except that request is specific to the members of an `enum` rather than the `enum` name itself.

I ask because I have some protobuf that has a fairly odd naming convention that _mostly_ follows the [Go style](https://google.github.io/styleguide/go/decisions#initialisms), where initialisms are generally upper-cased, but we have some words in our names that are kind of like "marketing names" where there's a specific capitalization (think `eXistenZ` but slightly less ludicrous). I think if we were designing our schema today we'd probably treat these as normal words (`existenz`/`Existenz`), but at this point it's probably too late to make that change without a lot of unnecessary code churn in projects that use the JSON serialization of the Proto structs.

So it would be very convenient for us to have a way to simply _turn off_ Prost's automatic renaming.

**Comment by stefanvanburen** [comment]
Have also seen an issue like this crop up with the following code ([BSR permalink](https://buf.build/cosmos/cosmos-sdk/file/8cb798d7bf0561e119ebba3019533573833238fa:cosmos/staking/v1beta1/authz.proto#L22)):

```proto
// NOTE: Trimmed out irrelevant context.
message StakeAuthorization {
  oneof validators {
    Validators allow_list = 2;
    Validators deny_list = 3;
  }
  message Validators {
    repeated string address = 1;
  }
}
```

prost will generate both a `pub struct Validators` for the nested message, and a `pub enum Validators` for the `oneof`, causing a conflict. Both the oneof and nested message names are following the style guide.

**Comment by caspermeijn** [comment]
> prost will generate both a `pub struct Validators` for the nested message, and a `pub enum Validators` for the `oneof`, causing a conflict. Both the oneof and nested message names are following the style guide.

I think this is a different and more interesting issue than the original example. I feel like in this case, the enum for `oneof` should be automatically renamed to prevent this conflict. In the C++ generated code there is no generated type for `oneof` and therefore it doesn't result in a conflict. 

@stefanvanburen Are you willing to work on a PR for solving this name conflict?

**Comment by stefanvanburen** [comment]
@caspermeijn I unfortunately don't have the bandwidth / Rust expertise to contribute a fix at the moment. For the future, do you have a suggestion as to how something like this would be implemented? I imagine prost has to deal with conflicting identifiers in other situations.

**Comment by caspermeijn** [comment]
> do you have a suggestion as to how something like this would be implemented? I imagine prost has to deal with conflicting identifiers in other situations.

The current situation is that it doesn't handle those cases. [Sometimes](https://github.com/tokio-rs/prost/blob/ad5650b0c96b7e5c1f8310462af7f6e35c2a0bfe/prost-build/src/code_generator.rs#L1182) it is detected, and a panic is thrown during build. If you follow protobuf naming convention and choose unique names, then there are no problems. This is the first case, that I have seen, where a design decision of `prost` is troublesome.

**Comment by caspermeijn** [comment]
The example of @stefanvanburen has an other issue dedicated to it: https://github.com/tokio-rs/prost/issues/505

**Comment by caspermeijn** [comment]
I am closing this issue.

In the original example, the names only differ by casing. That is a bad idea and I have no interest in supporting that.

The other examples are advised against in protobuf style guide. The reasoning in the style guide is exactly this kind of problem. I suggest following the style guide for naming.  https://protobuf.dev/programming-guides/style/
